lib/puppet_x/puppetlabs/powershell/powershell_manager.rb

```diff
@@ -395,17 +395,14 @@ def read_streams
           # read a Little Endian 32-bit integer for length of response
           expected_response_length = pipe.sysread(4).unpack('V').first
 
-          if expected_response_length == 0
-            nil
-          else
-            # reads the expected bytes as a binary string or fails
-            buffer = ""
-            # Reads in the pipe data 8K, or less, at a time
-            while (buffer.length < expected_response_length)
-              buffer << pipe.sysread([expected_response_length - buffer.length, 8192].min)
-            end
-            buffer
+          next nil if expected_response_length == 0
+          # reads the expected bytes as a binary string or fails
+          buffer = ""
+          # Read in the pipe data until complete
+          while (buffer.length < expected_response_length)
+            buffer.concat(pipe.sysread(expected_response_length - buffer.length))
           end
+          buffer
```

Comment:
Ok... so I think you're actually back to a more complicated form of square one here.  The original code before your changes for 8k was:

```ruby
return nil if expected_response_length == 0

# reads the expected bytes as a binary string or fails
pipe.sysread(expected_response_length)
```

In the new code, buffer starts at length 0, so let's fill in those values for the first iteration:

```
while (0 < expected_length)
  buffer.concat(pipe.sysread(expected_length - 0))
end
```

So it seems like you might not be correctly reproing the initial failure conditions you had if this works now...

Reply:
So as a nit, I should probably be using `do loop` not `while`.

the `concat` is accumulating the `sysread` calls. so the first call has zeros of course, but subsequent will have the buffer either full or partially full.

```
C:\Source\puppetlabs-powershell\powershell [fix-pipe-reader ≡]> irb
irb(main):001:0> buffer = ""
=> ""
irb(main):002:0> buffer.concat("Hello World")
=> "Hello World"
irb(main):003:0> buffer
=> "Hello World"
```

Reply:
I changed the `while` to `do`

Reply:
To follow up on this... the part that I was missing is that `sysread` reads "up to" the specified length, but no guarantees that it reads all.

So we need to keep track of what we've actually read vs what we expect to read.